Include `regex` when linting validators

davelopez opened the pull request:
I think this should catch invalid `regex` expressions too when linting validators.

See https://github.com/galaxyproject/galaxy/issues/16676#issuecomment-1717275714

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Include `regex` when linting validators

This will catch invalid regex expression too.

mvdbeek commented on lib/galaxy/tool_util/linters/inputs.py:
```diff
@@ -395,10 +395,10 @@ def lint_inputs(tool_source: "ToolSource", lint_ctx: "LintContext"):
                 if validator.text is None:
                     lint_ctx.error(
-                        f"Parameter [{param_name}]: expression validators are expected to contain text", node=validator
+                        f"Parameter [{param_name}]: {vtype} validators are expected to contain text", node=validator
```
I'm a little surprised about the existing expression linting ... `expression` is just any python expression, so I think the `re.compile(validator.text)` validation is wrong ?

davelopez replied:
Looking at the tool's source in IUC I thought exactly that, that `expression` here should have been `regex` from the beginning... but I was not sure how it was handled internally if those expression texts are then converted to matching regex somehow...

Should I just replace `expression` with `regex` and call this a bug?

mvdbeek replied:
I'd maybe turn line 403 into `elif vtype == "regex":`, the non-empty text linting is probably correct.

I assume this wasn't noticed because most expressions probably compile fine into a regex ?

davelopez replied:
> the non-empty text linting is probably correct

You mean a `regex` cannot be empty and must also have text right?

mvdbeek replied:
both `regex` and `expresison` should not be empty, iow `if validator.text is None:` should be applied to both

davelopez committed:
Add test for regex validator must contain text

davelopez committed:
Consider only `regex` as regular expression

`expression` is just a Python (eval) expression.

mvdbeek reviewed (approved):
Thank you!

bernt-matthias commented on test/unit/tool_util/test_tool_linters.py:
```diff
@@ -410,7 +410,7 @@
             <validator type="regex" filename="blah"/>
             <validator type="expression"/>
-            <validator type="expression">[</validator>
+            <validator type="regex">[</validator>
```
Maybe just add a validator here?

davelopez replied:
Do you mean just calling `eval` with the expression text?

bernt-matthias replied:
I see. The test was just wrong. My bad.

But we could really call something like `compile(source, filename)` or `ast.parse(expr, filename='<unknown>', mode='exec')` on the expression content.

davelopez replied:
Ahh sure, if there are ways to check the "syntax" without really evaluating I guess we can add it.
I'll investigate :+1:

davelopez replied:
I used `ast.parse` and added a couple of test lines. Let's see if this works for all expressions :crossed_fingers:

davelopez replied:
From the test failures, it seems different versions of Python return slightly different error messages for invalid expressions. I'll see if I can loosen the asserts a bit.

davelopez replied:
I have relaxed the LintMessage comparison to check if the string is contained in the actual message instead of checking for the exact thing, I don't like that much that is implemented in the `__eq__` function, but it is only used for testing and we should take into account that error messages tend to change or improve over time without needing to fail the test when they do.
If this is not what we want, I can try a different more ad-hoc approach.

davelopez committed:
Validate the syntax of the expressions

davelopez committed:
Add valid test expression

davelopez committed:
Relax lint message string comparison

This will allow to check that certain reasonable parts of the message are present without expecting an exact match.

This is useful, when the warning or error messages slightly differ from version to version.
